### PR TITLE
feat(portforwarding): Allow running script upon port forwarding success

### LINKED ISSUE #2351: Feature request: Webhook call when forwarded port change
### What's the feature 🧐

I use gluetun with PIA and forwarding. I have seen many ways to read the forwarded port from gluetun (most of them using the fwd_port file, wich I read is being deprecated) but I think the best way would be gluetun that makes a web request by itself calling a user-configured webhook when the port is being changed and maybe at the start of gluetun too.
Making the web request entirely configurable with a template inside the docker compose file (or a separate config file) would be ideal. 
I understand it can take some time to implement such a thing, so a simple GET or POST request with the forwarded port to a user-configurable endpoint it'd be nice too.

I found [a program (written in Go)](https://github.com/afiestas/gluetun-sync) that implement this, if you want to take a look 

Thank you for your work!

### Extra information and references

_No response_

### PR DESCRIPTION
**Pull request by lavalleeale**
This PR allows an environment variable (VPN_PORT_FORWARDING_STATUS_SCRIPT) to be set that will run a script whenever there is an update to port forwarding status that could be used to sync with other programs that need to know what ports have been forwarded.

EDIT by qdm12:

- Fixes #2351 
- Partly addresses #1785 
- May refer to #616, #1555 (more to completed)

### CODE REVIEW

File: internal/portforward/service/script.go
Comment (qdm12) [inline internal/portforward/service/script.go:16]: we should have this as a comma separated value ideally, it's easier to split by comma then taking a variable number of arguments I think (across shells)

File: internal/portforward/service/script.go
Comment (qdm12) [inline internal/portforward/service/script.go:23]: This should be context (ctx context.Context) dependent, I think there is a exec.CommandContext for that.

File: internal/portforward/service/start.go
Comment (qdm12) [inline internal/portforward/service/start.go:72]: This needs to change anyway, but just in case ™️ it should be

```suggestion
	if s.settings.Scriptpath != "" {
```

File: internal/portforward/service/start.go
Comment (qdm12) [inline internal/portforward/service/start.go:92]: nit unneeded change, please revert 😉

**Review by qdm12 — CHANGES_REQUESTED** [review]
Nice, thank you for this PR 👍 

We need to notably:

- change this to a run command (instead of script file)
- make code ready for a down command (later 🙏, if anyone needs such command)
- [make a PR](https://github.com/qdm12/gluetun-wiki/compare) to the github wiki to document this (probably in the setup/advance/vpn-port-forwarding document) and mention the command is a single command, which doesn't understand shell specific syntax such as `&&`, and one should use `/bin/sh -c "my shell syntax"` to do so if they want (bash isn't installed by default, only sh)

Thanks!

**Comment by lavalleeale** [comment]
Ok, I believe that I have resolved all the issues with the initial implementation, I am not a huge fan of comma separating the ports, but that is how the current code works to give maximum compatibility

File: internal/configuration/settings/portforward.go
Comment (qdm12) [inline internal/configuration/settings/portforward.go:36]: Sorry to re-iterate on this, can you rename this to `UpCommand`? To plan ahead for an eventual `DownCommand` 🤔 Thanks!

File: internal/portforward/service/command.go
Comment (qdm12) [inline internal/portforward/service/command.go:26]: This should be replaced by `github.com/qdm12/golibs/command` (soon to be local in #2418) and plugged in the logger somehow, to avoid race conditions on os.Stdout (and have smushed logs). I'll push this on your branch no worries😉

File: internal/portforward/service/command.go
Comment (qdm12) [inline internal/portforward/service/command.go:12]: any better suggestion than using CSV to seperate values? That could be added in #1113 !

File: internal/portforward/service/settings.go
Comment (qdm12) [inline internal/portforward/service/settings.go:15]: Same here, please rename to `UpCommand` 😉

File: internal/portforward/service/start.go
Comment (qdm12) [inline internal/portforward/service/start.go:77]: I wonder, if the command fails, should port forwarding fail to start, and possibly then crash the container 🤔 
Maybe it would make more sense to just log it as an error, especially given port forwarding is *working*, it's just the user provided script that failed (which is somehow likely to happen imo)

**Review by qdm12 — COMMENTED** [review]
Nice job! 👍 
Just 2 renames to do 😉 
And once done I'll handle the using-that-command-package-and-plug-lines-to-the-logger. I just worked on it yesterday so it's still fresh in my mind 😉 !

File: internal/configuration/settings/portforward.go
Comment (qdm12) [inline internal/configuration/settings/portforward.go:178]: Also please add `VPN_PORT_FORWARDING_UP_COMMAND=` to the Dockerfile's ENV list thanks!

File: internal/portforward/service/command.go
Comment (qdm12) [inline internal/portforward/service/command.go:11]: I'll add a test to make sure for example this works: `/bin/sh -c "echo {{PORTS}}"` - given it plugs arguments by splitting them by spaces, I'm not sure it would handle that case with double quotes correctly. Will check!

File: internal/configuration/settings/portforward.go
Comment (qdm12) [inline internal/configuration/settings/portforward.go:32]: Self comment to self-resolve: don't forget to merge PR https://github.com/qdm12/gluetun-wiki/pull/89 !

**Comment by qdm12** [comment]
Hi @lavalleeale could you give me (force) push access to your Gluetun branch `master`? I'm done with the changes, just waiting to force push them there.

**Comment by thibautd** [comment]
Thanks for working on this, can't wait !

**Comment by iceteaSA** [comment]
Anything happening here?

**Comment by qdm12** [comment]
@lavalleeale I still don't have push access to your fork, can you please allow force push access? Otherwise I'll create another PR with another branch, but would like to give you the credits for it ideally!

**Comment by lavalleeale** [comment]
Done!

**Comment by qdm12** [comment]
Thanks @lavalleeale 
Just waiting for CI to pass, and I'll re-recheck my own still unresolved comments to make I've done them (has been a while already!).

**Review by qdm12 — APPROVED** [review]

**Comment by qdm12** [comment]
Merged, thanks! I've tested it locally, it looks like it's working fine.
I also added VPN_PORT_FORWARDING_DOWN_COMMAND` in the next commit, which does what its name says.
And also pushed f9bdb219d0837b6d8efad38bab77657523428246 to support quoting better i.e. `echo "hello"` would previously be changed to `echo "hello` which would fail the parsing step, now it's fixed.

**Comment by KSmanis** [comment]
@qdm12 Thanks for merging! I've already successfully migrated my setup from polling the control server to this hook, but I'd rather not follow the master branch. Any plans on a new release soon?

**Comment by thibautd** [comment]
For those wanting to use this new feature with Transmission, I wrote a simple Python script to automatically update the `Peer Port` in Transmission. It's written in Python and uses the [uv](https://github.com/astral-sh/uv) package manager.

**docker-compose.yml**
```
services:
  gluetun:
    build: ./gluetun
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun:/dev/net/tun
    environment:
      - Your other env vars here...
      - VPN_PORT_FORWARDING_UP_COMMAND=/bin/uv run --script /transmission-sync.py {{PORTS}}
    restart: unless-stopped
    networks:
      - internal
  transmission:
    image: lscr.io/linuxserver/transmission:latest
    container_name: transmission
    network_mode: "service:gluetun"
    volumes:
      - # Your volumes here
    restart: unless-stopped
```

I had to create a custom Docker image for gluetun to install Python, `uv` and my script inside:

**gluetun/Dockerfile**
```
FROM qmcgaw/gluetun:latest

RUN apk add --no-cache python3

COPY --from=ghcr.io/astral-sh/uv:latest /uv /uvx /bin/
COPY transmission-sync.py /
```

Finally, the Python script wich uses https://pypi.org/project/transmission-rpc/.

**gluetun/transmission-sync.py**
```
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "requests",
#     "transmission-rpc",
# ]
# ///

import sys
import requests
from transmission_rpc import Client

PORTS = [int(p) for p in sys.argv[1].split(",")]

print('Set port: ', PORTS)

CLIENT = Client(host="127.0.0.1", port=9091)
CLIENT.set_session(peer_port=PORTS[0])

print('Done')
```

I put that here in case it can be useful :-)

**Comment by Marsu31** [comment]
Thank you for this great feature !!! :-D
My 50 cents, for qbittorrent container:

`VPN_PORT_FORWARDING_UP_COMMAND=wget -qO - --post-data "json={\"listen_port\":{{PORTS}}}" http://127.0.0.1:${QBITTORRENT_UI_PORT}/api/v2/app/setPreferences`

Prerequisites:
Whitelist localhost connection in qbittorent options.
Have environment variable `QBITTORRENT_UI_PORT` set or replace by port.

Have a great holiday :-)

**Comment by ramennbowls** [comment]
Thank you @Marsu31 for your command. I managed to simplify it like so 👇
```env
VPN_PORT_FORWARDING_UP_COMMAND=wget -qO- --post-data=json={\"listen_port\":{{PORTS}}} http://127.0.0.1:8080/api/v2/app/setPreferences
```
---
If your Gluetun container starts before qBittorrent, the command will fail because there is no service to contact yet. personally have qBittorrent depend on Gluetun, and will not start until it's healthy. 
```yml
qbittorrent:
    ...
    depends_on:
      gluetun:
        condition: service_healthy
```
For reliability, consider adding the `--retry-connrefused` and `--tries=6` flags.
- `--retry-connrefused` tells wget to retry if the address doesn't resolve (incorrect or likely down). Optionally, you can use the `--waitretry` flag to specify the time to wait in seconds between retries. By default, `--retry-connrefused` waits 1 second but doubles with each retry _(1s → 2s → 4s → 8s ...)_.
- `--tries` allows you to specify the maximum number of retries. I recommend setting this to 6, because `--retry-connrefused` doubles its wait between retries (unless specified otherwise) which totals to ~30 seconds worth of retry.

All in all, I'd suggest using this command 👇
```env
VPN_PORT_FORWARDING_UP_COMMAND=wget --retry-connrefused --tries=6 -qO- --post-data=json={\"listen_port\":{{PORTS}}} http://127.0.0.1:8080/api/v2/app/setPreferences
```
---
Personally, instead of using the `-qO-` flag, I prefer to use `-nv` which gives you a less verbose output so you know what's going on, but Gluetun seems to interpret all outputs from the command as errors so 🤔
```md
ERROR [port forwarding] failed: Connection refused.
ERROR [port forwarding] URL:http://127.0.0.1:8080/api/v2/app/setPreferences [0/0] -> "setPreferences" [4] *👈 not an error*
```

**Comment by redge76** [comment]
How do you guys handle the gluetun container restart ? 
If qbittorrent is connected to the gluetun network, it lose connectivity until it is restarted.
Some background: https://stackoverflow.com/questions/47088261/restarting-an-unhealthy-docker-container-based-on-healthcheck
(the second option does not work with qbittorent ...)

**Comment by Marsu31** [comment]
Thank you @ramennbowls for your improvements !  :-)

The lack of logging made me not see the error you pointed out.

To see output not in "ERROR" category I had to surround the command with shell and redirect stderr like this:

`VPN_PORT_FORWARDING_UP_COMMAND=/bin/sh -c '/usr/bin/wget -O- --retry-connrefused --post-data "json={\"listen_port\":{{PORTS}}}" http://127.0.0.1:${QBITTORRENT_UI_PORT}/api/v2/app/setPreferences 2>&1'`

I don't use `-nv` nor `-q` to be sure to see something. It doesn't happen often, also there is no performance problem (for me ;-) ).

**Pull request merged**
